[typescript] typescript client generator supports import mapping

uesyn opened the pull request:
<!-- Enter details of the change here. Include additional tests that have been done, reference to the issue for tracking, etc. -->

[typescript-node client generator](https://openapi-generator.tech/docs/generators/typescript-node) supports import mapping, but [typescript client generator](https://openapi-generator.tech/docs/generators/typescript/) doesn't.
This PR makes typescript client generator support import mapping feature.

I verified that import mapping works correctly in the following way.

- `./bin/generate-samples.sh` generates no changes because which doesn't use import mapping.
- Generate sample codes with import-mapping.

```sh
$ java -jar ./modules/openapi-generator-cli/target/openapi-generator-cli.jar generate -g typescript --import-mappings=Tag="../../types" -o out -i https://raw.githubusercontent.com/openapitools/openapi-generator/master/modules/openapi-generator/src/test/resources/3_0/petstore.yaml --additional-properties=useInversify=true

$ grep -1 -r "../../types" out
out/models/ObjectSerializer.ts-export * from './Pet';
out/models/ObjectSerializer.ts:export * from '../../types';
out/models/ObjectSerializer.ts-export * from './User';
--
out/models/ObjectSerializer.ts-import { Pet     , PetStatusEnum   } from './Pet';
out/models/ObjectSerializer.ts:import { Tag } from '../../types';
out/models/ObjectSerializer.ts-import { User } from './User';
--
out/models/all.ts-export * from './Pet'
out/models/all.ts:export * from '../../types'
out/models/all.ts-export * from './User'
--
out/models/Pet.ts-import { Category } from './Category';
out/models/Pet.ts:import { Tag } from '../../types';
out/models/Pet.ts-import { HttpFile } from '../http/http';
--
out/services/ObjectParamAPI.ts-import type { Pet } from '../models/Pet';
out/services/ObjectParamAPI.ts:import type { Tag } from '../../types';
out/services/ObjectParamAPI.ts-import type { User } from '../models/User';
--
out/services/ObservableAPI.ts-import { Pet } from "../models/Pet";
out/services/ObservableAPI.ts:import { Tag } from "../../types";
out/services/ObservableAPI.ts-import { User } from "../models/User";
--
out/services/PromiseAPI.ts-import { Pet } from "../models/Pet";
out/services/PromiseAPI.ts:import { Tag } from "../../types";
out/services/PromiseAPI.ts-import { User } from "../models/User";
--
out/types/ObjectParamAPI.ts-import { Pet } from '../models/Pet';
out/types/ObjectParamAPI.ts:import { Tag } from '../../types';
out/types/ObjectParamAPI.ts-import { User } from '../models/User';
--
out/types/ObservableAPI.ts-import { Pet } from '../models/Pet';
out/types/ObservableAPI.ts:import { Tag } from '../../types';
out/types/ObservableAPI.ts-import { User } from '../models/User';
--
out/types/PromiseAPI.ts-import { Pet } from '../models/Pet';
out/types/PromiseAPI.ts:import { Tag } from '../../types';
out/types/PromiseAPI.ts-import { User } from '../models/User';
```

close: https://github.com/OpenAPITools/openapi-generator/issues/12907

<!-- Please check the completed items below -->
### PR checklist
 
- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [x] Run the following to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) and update samples:
  ```
  ./mvnw clean package 
  ./bin/generate-samples.sh
  ./bin/utils/export_docs_generators.sh
  ``` 
  Commit all changed files. 
  This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. 
  These must match the expectations made by your contribution. 
  You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/configs/java*`. 
  For Windows users, please run the script in [Git BASH](https://gitforwindows.org/).
- [x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master` (6.1.0) (minor release - breaking changes with fallbacks), `7.0.x` (breaking changes without fallbacks)
- [x] If your PR is targeting a particular programming language, @mention the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) members, so they are more likely to review the pull request.

@TiFu (2017/07) @taxpon (2017/07) @sebastianhaas (2017/07) @kenisteward (2017/07) @Vrolijkx (2017/09) @macjohnny (2018/01) @topce (2018/10) @akehir (2019/07) @petejohansonxo (2019/11) @amakhrov (2020/02) @davidgamero (2022/03) @mkusaka (2022/04)

uesyn commented:
`./bin/generate-samples.sh` and `./bin/utils/export_docs_generators.sh` did not change anythin.

macjohnny commented:
@bodograumann can you review this?

bodograumann commented:
First of all, I am very happy that you decided to contribute, @uesyn!

Code change looks good so far.
I am just not completely sure what the scope of the feature should be.
Is there some documentation for it? All I could find was https://openapi-generator.tech/docs/customization#bringing-your-own-models which is very specific to java.

One thing I was wondering about in particular are absolute imports. Afaics, currently only relative paths would be supported.

uesyn commented:
@bodograumann

Thank you for the quick review!

> I am just not completely sure what the scope of the feature should be.
> Is there some documentation for it?  All I could find was https://openapi-generator.tech/docs/customization#bringing-your-own-models which is very specific to java.

I coudn't find other than that docs too.
The docs example code is specific to java, but import mapping feature is needed by many languages in common.
The docs said "When doing this, every location that references a specific model will refer back to your classes. Note, this may not apply to all languages...", I understood it that all languages is not implemented import mapping.
In fact, typescript-node supports import mapping, and I found that [kubernetes/gen](https://github.com/kubernetes-client/gen/blob/f6aa01426d0a0efaf0692db66952ff79a8cd21c6/openapi/typescript.xml#L25-L27) uses it to treat external objects in generated codes.

typescript-node generator uses [deprecated request module](https://www.npmjs.com/package/request).
Therefore, I think some users consider to migrate from typescript-node to typescript client generator.
To migrate smoothly for the users, I think typescript client generator should support import mapping same as typescript-node. As far as I confirmed, kubernetes/gen is one of them.

> One thing I was wondering about in particular are absolute imports. Afaics, currently only relative paths would be supported.

Although I uses relative path for import mapping in my example, [absolute imports](https://create-react-app.dev/docs/importing-a-component/#absolute-imports) is supported too in this PR:smile:.

```
$ java -jar ./modules/openapi-generator-cli/target/openapi-generator-cli.jar generate -g typescript --import-mappings=Tag="components/Order" -o out -i https://raw.githubusercontent.com/openapitools/openapi-generator/master/modules/openapi-generator/src/test/resources/3_0/petstore.yaml --additional-properties=useInversify=true
$ grep -1 -r "components/Order" out
out/models/ObjectSerializer.ts-export * from './Pet';
out/models/ObjectSerializer.ts:export * from 'components/Order';
out/models/ObjectSerializer.ts-export * from './User';
--
out/models/ObjectSerializer.ts-import { Pet     , PetStatusEnum   } from './Pet';
out/models/ObjectSerializer.ts:import { Tag } from 'components/Order';
out/models/ObjectSerializer.ts-import { User } from './User';
--
out/models/all.ts-export * from './Pet'
out/models/all.ts:export * from 'components/Order'
out/models/all.ts-export * from './User'
--
out/models/Pet.ts-import { Category } from './Category';
out/models/Pet.ts:import { Tag } from 'components/Order';
out/models/Pet.ts-import { HttpFile } from '../http/http';
--
out/services/ObjectParamAPI.ts-import type { Pet } from '../models/Pet';
out/services/ObjectParamAPI.ts:import type { Tag } from 'components/Order';
out/services/ObjectParamAPI.ts-import type { User } from '../models/User';
--
out/services/ObservableAPI.ts-import { Pet } from "../models/Pet";
out/services/ObservableAPI.ts:import { Tag } from "components/Order";
out/services/ObservableAPI.ts-import { User } from "../models/User";
--
out/services/PromiseAPI.ts-import { Pet } from "../models/Pet";
out/services/PromiseAPI.ts:import { Tag } from "components/Order";
out/services/PromiseAPI.ts-import { User } from "../models/User";
--
out/types/ObjectParamAPI.ts-import { Pet } from '../models/Pet';
out/types/ObjectParamAPI.ts:import { Tag } from 'components/Order';
out/types/ObjectParamAPI.ts-import { User } from '../models/User';
--
out/types/ObservableAPI.ts-import { Pet } from '../models/Pet';
out/types/ObservableAPI.ts:import { Tag } from 'components/Order';
out/types/ObservableAPI.ts-import { User } from '../models/User';
--
out/types/PromiseAPI.ts-import { Pet } from '../models/Pet';
out/types/PromiseAPI.ts:import { Tag } from 'components/Order';
out/types/PromiseAPI.ts-import { User } from '../models/User';
```

I'm not good at English speaker, so if I coudn't understand what you mean, please tell me again. :pray:

davidgamero commented:
Hello @uesyn you are absolutely correct about the input mapping usage for kubernetes/gen. I am currently migrating it use the new generator for the exact reason you mentioned (request lib deprecation) and this would be very helpful. For now I just have been working around it as I complete the work on migrating auth

bodograumann commented on modules/openapi-generator/src/main/resources/typescript/types/ObjectParamAPI.mustache:
```diff
@@ -7,7 +7,7 @@ import { Observable } from 'rxjs';
 {{#models}}
 {{#model}}
-import { {{{ classname }}} } from '../models/{{{ classFilename }}}{{extensionForDeno}}';
+import { {{{ classname }}} } from '{{{ importPath }}}{{extensionForDeno}}';
```
There are still still some imports in the `types/*` templates, which explicitly reference the `models` directory, instead of taking it from the `importPath`.
I think we don't actually need those imports anymore. Can you please remove the lines of the form
```
import * as models from '../models/all{{extensionForDeno}}';
```
from the three files?

uesyn replied:
hmm... It’s exactly as you said.
Remove lines from three files.

bodograumann reviewed (changes requested):
I think we should try to use the existing machinery for import mappings.
Have a look at `DefaultGenerator.getAllImportsMappings`: https://github.com/OpenAPITools/openapi-generator/blob/master/modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultGenerator.java#L1244-L1255

In particular it is also necessary to remove the following from `TypeScriptClientCodegen` then:
```
        // clear import mapping (from default generator) as TS does not use it
        // at the moment
        importMapping.clear();
```

uesyn commented on modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/TypeScriptClientCodegen.java:
```diff
@@ -809,7 +819,6 @@ public void processOpts() {
 
         // change package names
         apiPackage = this.apiPackage + ".apis";
-        modelPackage = this.modelPackage + ".models";
```
This value is set in https://github.com/OpenAPITools/openapi-generator/pull/12957/files#diff-3e73b8c7b3ba091813977ee07513b9da4e90b365475e91f4f714beb98ffb736dR240.

uesyn commented on modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/TypeScriptClientCodegen.java:
```diff
@@ -325,8 +325,7 @@ public OperationsMap postProcessOperationsWithModels(OperationsMap operations, L
         List<Map<String, String>> imports = operations.getImports();
         for (Map<String, String> im : imports) {
-            im.put("filename", im.get("import").replace(".", "/"));
-            im.put("classname", getModelnameFromModelFilename(im.get("import")));
```
If my understanding is correct, `classname` is set in https://github.com/OpenAPITools/openapi-generator/blob/master/modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultGenerator.java#L1272, and that `value` should equal getModelnameFromModelFilename(im.get("import")).

uesyn commented on modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/TypeScriptClientCodegen.java:
```diff
@@ -731,13 +734,20 @@ private List<Map<String, String>> toTsImports(CodegenModel cm, Set<String> impor
+        if (importMapping.containsKey(className)) {
+            return importMapping.get(className);
+        }
+	return filename;
```
fixed👍

uesyn commented:
@bodograumann

Thank you for your review!
Fixed the codes you pointed out.

In addtion that, I removed the lines containing `importMapping.clear();`.
I read [the `DefaultGenerator.getAllImportsMappings` codes](https://github.com/OpenAPITools/openapi-generator/blob/master/modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultGenerator.java#L1244-L1255), but I'm sorry that I couldn't understand what you would like to tell me. 
I think we have already been using the machinery, if my understanding is correct. (but I'm not famillier with openapi generator implementation)
If you have some concrete improvements in addition to removing the lines containing `importMapping.clear();`, please tell me. :pray:

bodograumann commented:
Very nice :rocket: 
I am only starting to understand this particular logic in `DefaultGenerator` myself.
So it is also not quite clear to me, what the best solution will be at the end,
but it seems to me, that we could possibly simplify the code in the typescript generator part,
when we reuse what is already there in the generic code.

I will have a more thorough look and let you know what I find out. :-)

uesyn commented:
Rebased to resolve conflicts.

macjohnny commented:
@bodograumann please let me know when this PR is ready to be merged.

bodograumann commented:
Before I go further down the rabbit hole: what do you think about the approach of using an `ImportMap` class, which stores the models that are imported from each source, not only a list of import sources?
Cf this commit: https://github.com/bodograumann/openapi-generator/commit/61476b07501c45d18a8dac649332c12d016b58af
In particular these changes in `DefaultGenerator`, `typescript/model.musache` and the new class `models.ImportMap`

I feel like this could be useful in other languages, other than typescript, as well. E.g. python has these kinds of imports in principal.
Maybe we could add another constructor for the case where only a single model is imported from some source with the `ImportMap(String importPath, String importModel)` signature. For now I simply left away the models in most places, because they are only used in the changes I mode to the typescript template here (for now).

@macjohnny, @wing328 , @spacether

macjohnny commented:
@amakhrov @akehir  any opinion here?

uesyn commented:
@bodograumann 
You have nice idea :+1:
However, In this PR, we should focus only the feature to introduce import mapping to typescript client generator.
I will be happy to modify the PR in case that it is not good.
However, if we should refactor the code related to import mapping for all languages, I think we should make another PR.

bodograumann commented:
You are probably right, @uesyn . I will gladly open a separate PR for it later. Though I hope to at least get some feedback whether this change has a chance of being included. There is also more work to be done there, e.g. I would harmonize import handling for operations and models.

For now I have added two commits here: https://github.com/bodograumann/openapi-generator/tree/ts-cleaner-importmap
That is what I meant by reducing specialized logic. `DefaultGenerator` already does the import mapping. All we need to do, is include it in our custom `toTsImports` as well.
Another test that essentially does `DefaultGenerator.processModels` in the typescript context would be nice, but I don't see how to do that without much effort (like creating a whole new sample directory).

Side note: what I now also noticed, is that our `toTsImports` is essentially `getAllImportsMappings` and `toImportsObjects` from `DefaultGenerator`. So if we don't try to group classes by import path, we might find an quicker solution than what I proposed above. (In another pr though :-D )

uesyn commented:
@bodograumann 
I checked your great two commits.
Your two additional commits look good if we can tolerate redundant  import path in generated codes. :+1:
In my local tests with import mapping, these worked well.
I can cherry-pick them, so if you are ok, I would like do so.

bodograumann commented:
That would be fantastic, @uesyn.
I think these kind of paths are perfectly fine in generated code and even allow moving the source files between sibling folders without breaking anything.

bodograumann commented:
I think we are ready to merge now, @macjohnny .
